processor/attributesprocessor/attributes.go

```diff
@@ -126,39 +101,61 @@ func (a *attributesProcessor) Shutdown(context.Context) error {
 	return nil
 }
 
-func insertAttribute(action attributeAction, attributesMap map[string]*tracepb.AttributeValue) {
-	// Insert is only performed when the target key does not already exist
-	// in the attribute map.
-	if _, exists := attributesMap[action.Key]; exists {
+func (a *attributesProcessor) processSpan(span data.Span, serviceName string) {
+	if span.IsNil() {
+		// Do not create empty spans just to add attributes
 		return
 	}
 
-	setAttribute(action, attributesMap)
-}
-
-func updateAttribute(action attributeAction, attributesMap map[string]*tracepb.AttributeValue) {
-	// Update is only performed when the target key already exists in
-	// the attribute map.
-	if _, exists := attributesMap[action.Key]; !exists {
+	if a.skipSpan(span, serviceName) {
 		return
 	}
 
-	setAttribute(action, attributesMap)
+	attrs := span.Attributes()
+	for _, action := range a.config.actions {
+		// TODO https://github.com/open-telemetry/opentelemetry-collector/issues/296
+		// Do benchmark testing between having action be of type string vs integer.
+		// The reason is attributes processor will most likely be commonly used
+		// and could impact performance.
+		switch action.Action {
+		case DELETE:
+			attrs.Delete(action.Key)
+		case INSERT:
+			av, found := getAttributeValue(action, attrs)
+			if !found {
+				continue
+			}
+			attrs.Insert(action.Key, av)
+		case UPDATE:
+			av, found := getAttributeValue(action, attrs)
+			if !found {
+				continue
+			}
+			attrs.Update(action.Key, av)
+		case UPSERT:
+			av, found := getAttributeValue(action, attrs)
+			if !found {
+				continue
+			}
+			attrs.Upsert(action.Key, av)
+		case HASH:
+			hashAttribute(action, attrs)
+		}
+	}
 }
 
-func setAttribute(action attributeAction, attributesMap map[string]*tracepb.AttributeValue) {
+func getAttributeValue(action attributeAction, attrs data.AttributeMap) (data.AttributeValue, bool) {
 	// Set the key with a value from the configuration.
-	if action.AttributeValue != nil {
-		attributesMap[action.Key] = action.AttributeValue
-	} else if value, fromAttributeExists := attributesMap[action.FromAttribute]; fromAttributeExists {
-		// Set the key with a value from another attribute, if it exists.
-		attributesMap[action.Key] = value
+	if !action.AttributeValue.IsNil() {
```

Comment:
This needs to be done differently after merging https://github.com/open-telemetry/opentelemetry-collector/pull/809/files

Reply:
Done.